certbot/certbot
# certbot-auto-cron [#3522]

For #3522: modifies certbot-auto to help people setup renewal cron jobs (for Debian, openSUSE, and Red Hat).

High-level summary of how this works:
  - If certbot-auto is run in interactive mode (the default), the user is asked if they want a renewal cron job installed.
    - If run in non-interactive mode, don't install a cron jobs because we can't prompt.
  - If there is an existing root crontab:
    - Check if it already contains `certbot -q renew # generated by certbot`, and if so, skip adding another.
    - A backup is made at `/opt/eff.org/certbot/crontab.bak`.
    - Finally, modify the existing crontab by appending a line that calls `certbot -q renew` at a random time during the first half of the day, then at +12 hours after. The randomized time is setup using an offset via `sleep`.
  - If there's no existing root crontab, create one and add the same as above.
  - [Debian-specific] if `/etc/cron.d/certbot` exists (generated by Debian's packaged certbot), do not add anything to the root crontab. (This was requested by our Debian packager.)
  
  
Testing done:
  - Automated tests that were written (in `tests/letstest/scripts/test_certbot_auto_cron.sh` and `tests/letstest/scripts/test_certbot_auto_cron_debian_pkg.sh`):
    - [Debian-specific] check that we don't to add a crontab if `/etc/cron.d/crontab` exists
    - Check that backups of existing root crontabs are created.
    - Check that if no root crontab exists, we can create one.
    - Check that if a root crontab exists, we can append to the end of it without altering the existing contents.
    - Check that if someone runs certbot-auto twice, they don't end up with multiple renewal cron jobs being generated.
  - Miscellaneous notes:
    - OpenSUSE is not currently included in the AWS test environments.
        - If it is added, the last test would be expected to fail, as the output format of `crontab -l` happens to be different on openSUSE than on other OSs.
    - The tests run in non-interactive mode by default, so they rely on the `--force-cron` flag to trigger the relevant behaviour in certbot-auto-cron.
    - The tests are split into two files because the Debian-specific test required capturing stdout from the letsencrypt command, which seemed to break whenever the boulder instances responded with rate-limit errors. There's probably a simple way to fix this.
    - I'm aware that the tests fail in the first Ubuntu env specified in `targets.yaml`, but I've not seen any of the other tests pass on that, so I'm led to believe there's just something wrong with that environment.

  - Manual testing done on (OpenSUSE Tumbleweed, Ubuntu Trusty, and Fedora Rawhide):
    - Can install a crontab properly
    - Backup is made + we can append to existing crontab properly
    - Running it twice doesn't end in multiple lines being added
    - That if the user says they don't want a crontab, one isn't installed. (It's not possible to write an automated test for this case.)
  - Note: if you are testing this manually, you need to rename the file to `certbot-auto`, because when when it's run as `letsencrypt-auto`, it's in non-interactive mode by default for backwards-compatibility reasons.
